src/Processors/Transforms/WindowTransform.cpp

```diff
@@ -1970,6 +1969,121 @@ struct WindowFunctionRowNumber final : public WindowFunction
     }
 };
 
+// Usage: ntile(n). n is the number of buckets.
+struct WindowFunctionNtile final : public WindowFunction
+{
+    WindowFunctionNtile(const std::string & name_,
+            const DataTypes & argument_types_, const Array & parameters_)
+        : WindowFunction(name_, argument_types_, parameters_, std::make_shared<DataTypeUInt64>())
+    {
+        if (argument_types.size() != 1)
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Function {} takes exactly one parameter", name_);
+        }
+    }
+
+    bool allocatesMemoryInArena() const override { return false; }
+
+    void windowInsertResultInto(const WindowTransform * transform,
+        size_t function_index) override
+    {
+        if (!buckets) [[unlikely]]
+        {
+            checkWindowFrameType(transform);
+            const auto & current_block = transform->blockAt(transform->current_row);
+            const auto & workspace = transform->workspaces[function_index];
+            auto n = (*current_block.input_columns[
+                    workspace.argument_column_indices[0]])[
+                        transform->current_row.row].get<Int64>();
+            if (n <= 0)
+            {
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "ntile's argument must > 0");
+            }
+            buckets = static_cast<UInt64>(n);
```

Comment:
If we cant to `UInt64`, let's forbid bigger argument types (`(U)Int128/256`), `Decimals`, and `Float`. Or better just allow only `(U)Int8/16/32/64`